### ISSUE TITLE
Command line interface

### ISSUE THREAD

**Issue by abnerlee** [typora/typora-issues#1999]
Use command line to let Typora:

- Export file #715
- Open file with options
   - Open files in one tab #979
   -  Create file if not exist #721

**Comment by nilsj** [comment]
One slightly strange use case of mine of "Create file if not exists" i'd like to add to consideration:
Do not create the new file directly on disk. Only create a buffer (or whatever the typora equivalent is), and only create the file if it is saved.
I use this with sublime text to create possibly temporary notes via a keyboard shortcut.
When i want to save the note, I have already specified a name programmatically so I don't have to bother giving it a name.
I'd love to replace sublime text with typora for my notes.

**Comment by GiovanH** [comment]
One note about this: the main reason I need CLI support is to integrate Typora's non-markdown features, like mermaid charts. Ideally, a simple way to programatically replicate the "File->Export->HTML" option.

**Comment by zerostone** [comment]
Desire the command line export feature

**Comment by susu3621** [comment]
Me too, desire the command line export feature.
It's awesome tool, we are using it to replace MS word :-), for document automation.

**Comment by gudonglaile** [comment]
Open the file and jump to a specific line number, like Notepad++ does:
     start notepad++ "some file.txt" -n100 -c8
     start typora "some file.md" -n100 -c8
It is extremly useful to be called by text file searching tools like FileLocator!!!

**Comment by GiovanH** [comment]
I'm very disappointed that there's been little to no official follow up on this issue for almost two years. As I've mentioned, this is not just an annoyance or a nice feature request: not having these features exposed in a practical way (and the absence of CLI options for conversion and processing) makes Typora markdown files nonstandard and non-portable, which really defeats the purpose.

I'm disappointed and frustrated, and I'm going to have to look into converting all my markdown documents into a standard, useful form. With no support or even a commitment of support, I can no longer even recommend Typora.

**Comment by AlexanderMattheis** [comment]
Yeah that would be one of the greatest features ever. I've implemented once a `Powershell`-script to export my Word documents to PDF:

https://github.com/AlexanderMattheis/relationship-enhancement-scripts/blob/master/src/RecursiveDocToPdfConverter.ps1

Now, I want write me a a Typora-Bash script, since I've switched to Linux Mint.

**Comment by Mr-Wiredancer** [comment]
I traced all issues about CLI export option to this one. Is there any followup or any final decision on whether it will be implemented?

**Comment by yaakovazat** [comment]
> Use command line to let Typora:
> 
> * Export file #715
> * Open files in one window (multi-tabs) #979
> * Create file if not exist #721
> * Option for keep foreground (e.g: support edit git commit message) #649

If the Typora CLI comes up, that could be a really nice feature to use, thanks a lot !!!!!

**Comment by maxwellainatchi** [comment]
@abnerlee is there any followup on this? I want to be able to use Typora's HTML generation to publish web docs, but without a way to build from the command line, it's essentially useless.

**Comment by v01pe** [comment]
As Typora seems to support document conversion through Pandoc, as a workaround for conversion, you can use Pandoc directly: https://pandoc.org/MANUAL.html
For me it was already installed and is accessible through the Terminal (on macOS), I guess because it's installed via Typora? If that's not the case, you can install it directly.

**Comment by jochenw** [comment]
Would be nice to have an update on the status, please?

**Comment by PegasisForever** [comment]
I'm currently working on creating an unofficial exporter for typora, it aims to have the exact same output as typora, with some more options for advanced usage. It is written in nodejs (so I can use the same library as typora uses). However it is still in early stages, I'll put it on Github and share it here when I finish basic functions.

I suspect it'll take me 1-2 months from now to release the first version (though it's nothing compared to this 3-years-old issue), be patient!:)

**Comment by v01pe** [comment]
> Would be nice to have an update on the status, please?

Depends on for what you're here (this issue seems to be very multi-purpose). I recently discovered, there the **are** new export feature, one being a custom export option, where you can supply your own export scripts, etc. Also existing ones can be customized there – see the [custom export feature documentation](https://support.typora.io/Export/#other-formats-using-custom-commands). This fulfils my needs, others might have different requirements.

**Comment by PegasisForever** [comment]
Hi, just a status update on my typora cli. I know I'm behind the schedule :) I've finished all the architectural and algorithm heavy lifting stuff, the rest is pretty trival. Expect to see the first version in 2 weeks!

**Comment by YalandHong** [comment]
> I'm very disappointed that there's been little to no official follow up on this issue for almost ~two~ (edit: three) years. As I've mentioned, this is not just an annoyance or a nice feature request: not having these features exposed in a practical way (and the absence of CLI options for conversion and processing) makes Typora markdown files nonstandard and non-portable, which really defeats the purpose.
> 
> I'm disappointed and frustrated, and I'm going to have to look into converting all my markdown documents into a standard, useful form. With no support or even a commitment of support, I can no longer even recommend Typora.

I couldn't agree more. Currently Typora is not completely compatible with GitHub Flavored Markdown (#1789), making it a world of pain to publish a document written in Typora to my blog.

**Comment by scruel** [comment]
Hope you can support "export file via command line" feature soon!

**Comment by GiovanH** [comment]
They're not going to. They've decided to make this hostile software. Stop using it.

**Comment by WayneYam** [comment]
Does anyone know how to export markdown to PDF in command line that looks exactly like Typora's? I really love its style

**Comment by ik631** [comment]
@abnerlee when will you add command line support for md to pdf on Linux?

**Comment by mattiasJohnson** [comment]
@WayneYam Not a complete answer but if you use [PegasisForever/typora-parser](https://github.com/PegasisForever/typora-parser) from above in this thread you get html that is formatted exactly the same and then for going html->pdf you could probably use pandoc

**Comment by yingbo** [comment]
It will be great to have this feature. Currently I'm writing md using Typora, export to PDF to share the document. Would like to make the pdf generation as a part of CI/CD to automatically update the PDF file. I tried the [PegasisForever/typora-parser](https://github.com/PegasisForever/typora-parser) above, which has some differences as the vanilla Typora, especially the mathjs part, e.g. formula number, references, and etc. I feel it's great to be able to use the Typora cli directly. Thanks!

**Comment by cyrilDamae** [comment]
Generating pdf with typora from command line would be amazing for our CI too.
Can't wait to see this feature coming !

**Comment by YDX-2147483647** [comment]
> as a part of CI/CD

In my humble opinion, using Typora in CI/CD is a separate and more complicated issue. How to pass the license code to activate? What about the 3-device limit? Can other people trigger the CI (e.g. in PR)? Maybe [self-hosted runners](https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/about-self-hosted-runners) are required.

**Comment by cyrilDamae** [comment]
We precisely have self-hosted runners, our documentary environment is heavy because of CE and FDA appliances and the simple operation of exporting each document in pdf is painful. it would be so nice to have a CI exporting it on each release or on each merge request.

**Comment by sunshaoce** [comment]
I'm tired of feeling like an idiot, having to click a bunch of times just to export my Markdown files to HTML. We just need a single interface since most of the work is already done. PLEEEEEEEEEEASE let me use CI to export my files. I know Pandoc can export a simple HTML, but compared to Typora, it's like going out without wearing pants.

**Comment by cyrilDamae** [comment]
> > as a part of CI/CD
> 
> In my humble opinion, using Typora in CI/CD is a separate and more complicated issue. How to pass the license code to activate? What about the 3-device limit? Can other people trigger the CI (e.g. in PR)? Maybe [self-hosted runners](https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/about-self-hosted-runners) are required.

To return to the question of implementation in CI, a simple implementation would be to allow command line with one user license (one of the bundle) the license could be dedicated to a runner (bind to hardware for example). This is not a problem to have only one runner able to process documentation for several pipelines while many others could handle the heavier tasks. More importantly even for "manual" launch of pdf export by a specific user would be helpful to have a script able to launch many of them in a row.

**Comment by yanghanlin** [comment]
It's been 6+ years, and we are still waiting for this to happen.

**Comment by trojblue** [comment]
Any updates?
Or guess we'd better vibe code an actual solution somehow, using Typora css.
